src/accountsdb/account_store.zig

```diff
@@ -0,0 +1,211 @@
+const sig = @import("../sig.zig");
+const accounts_db = @import("lib.zig");
+
+const Allocator = std.mem.Allocator;
+
+const RwMux = sig.sync.RwMux;
+
+const Account = sig.core.Account;
+const Ancestors = sig.core.status_cache.Ancestors;
+const Pubkey = sig.core.Pubkey;
+const Slot = sig.core.Slot;
+
+const AccountSharedData = sig.runtime.AccountSharedData;
+
+const AccountsDB = accounts_db.AccountsDB;
+
+/// Interface for only reading accounts
+pub const AccountReader = union(enum) {
+    accounts_db: *AccountsDB,
+    thread_safe_map: *ThreadSafeAccountMap,
+    noop,
+
+    /// use this to deinit accounts returned by get methods
+    pub fn allocator(self: AccountReader) Allocator {
+        return switch (self) {
+            .noop => sig.utils.allocators.failing.allocator(.{}),
+            inline else => |item| item.allocator,
+        };
+    }
+
+    pub fn get(self: AccountReader, address: Pubkey, ancestors: *const Ancestors) !?Account {
+        return switch (self) {
+            .accounts_db => |db| try db.getAccount(&address), // TODO: PR #796
+            .thread_safe_map => |map| try map.get(address, ancestors),
+            .noop => null,
+        };
+    }
+
+    pub fn getLatest(self: AccountReader, address: Pubkey) !?Account {
+        return switch (self) {
+            .accounts_db => |db| try db.getAccount(&address),
+            .thread_safe_map => |map| try map.getLatest(address),
+            .noop => null,
+        };
+    }
+};
+
+/// Interface for both reading and writing accounts.
+///
+/// Do not use this unless you need to *write* into the database generically.
+/// Otherwise use AccountReader if you only need to read accounts.
+pub const AccountStore = union(enum) {
+    accounts_db: *AccountsDB,
+    thread_safe_map: *ThreadSafeAccountMap,
+    noop,
+
+    pub fn accountReader(self: AccountStore) AccountReader {
+        return switch (self) {
+            .accounts_db => |db| .{ .accounts_db = db },
+            .thread_safe_map => |map| .{ .thread_safe_map = map },
+            .noop => .noop,
+        };
+    }
+
+    /// use this to deinit accounts returned by get methods
+    pub fn allocator(self: AccountStore) Allocator {
+        return switch (self) {
+            .noop => sig.utils.allocators.failing.allocator(.{}),
+            inline else => |item| item.allocator,
+        };
+    }
+
+    pub fn get(self: AccountStore, address: Pubkey, ancestors: *const Ancestors) !?Account {
+        return switch (self) {
+            .accounts_db => |db| try db.getAccount(&address), // TODO: PR #796
+            .thread_safe_map => |map| try map.get(address, ancestors),
+            .noop => null,
+        };
+    }
+
+    pub fn getLatest(self: AccountStore, address: Pubkey) !?Account {
+        return switch (self) {
+            .accounts_db => |db| try db.getAccount(&address),
+            .thread_safe_map => |map| try map.getLatest(address),
+            .noop => null,
+        };
+    }
```

Comment:
Any reason these are duplicated in both this and the reader? I think we could delete these; reading from the store could be accomplished with `store.accountReader().get()`.